packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,489 @@
+   */
+  const queryAbci = async (
+    path,
+    { height = 0n, kind = PATHS.CHILDREN } = {},
+  ) => {
+    await null;
+
+    const codec = codecs[kind];
+    /** @type {Response|undefined} */
+    let response;
+    const route = createVstorageAbciRoute(height, kind, path);
+    const rpcAddress = getRpcAddress();
+
+    const url = rpcAddress + route;
+
+    try {
+      response = await fetch(url, { keepalive: true });
+    } catch (err) {
+      console.error(err);
+      throw createError(err.message, height, kind, path, rpcAddress);
+    }
+
+    if (!response.ok)
+      throw createError(await response.text(), height, kind, path, rpcAddress);
+
+    const { result } = /** @type {AbciResponse} */ (await response.json());
+
+    if (result.response?.code !== 0)
+      throw createError(
+        `Error code '${result.response.code}', Error message: '${result.response.log}'`,
+        height,
+        kind,
+        path,
+        rpcAddress,
+      );
+
+    return codec.response.decode(decodeBase64(result.response.value));
+  };
+
+  return { getLatestHeight, queryAbci };
+};
+
+/**
+ * @template {any} T
+ * @param {Array<T>} values
+ */
+const makeRoundRobin = values => {
+  let currentIndex = 0;
+
+  return () => {
+    const element = values[currentIndex];
+    currentIndex = (currentIndex + 1) % values.length;
+    return element;
+  };
+};
+
+/**
+ * @template {any} T
+ * @param {{queryClient: ReturnType<typeof makeQueryClient>}} powers
+ * @param {string} path
+ * @param {Partial<{ compat: boolean }>} [options]
+ * @returns {StreamTopic<T>}
+ */
+const makeStreamTopic = ({ queryClient }, path, options) => {
+  /**
+   * @param {Height} blockHeight
+   * @param {StreamCell['values']} values
+   */
+  function* allValuesFromCell(blockHeight, values) {
+    for (const value of values.reverse())
+      yield /** @type {Update<T>} */ ({
+        blockHeight,
+        value: JSON.parse(value),
+      });
+  }
+
+  /**
+   * @param {boolean} compatMode
+   * @param {ReturnType<typeof parseValue>} value
+   */
+  const assertValueStructure = (compatMode, value) => {
+    if (!(isStreamCell(value) || compatMode))
+      throw Error(`Expected a stream cell, got ${quote(value)}`);
+  };
+
+  /**
+   * @param {Height} [minimum]
+   * @param {Height} [maximum]
+   */
+  async function* iterate(minimum, maximum) {
+    /** @type {Array<Update<T>>} */
+    let values = [];
+
+    for await (const value of reverseIterate(maximum, minimum, true))
+      values = [value, ...values];
+
+    for (const value of values) {
+      assertValueStructure(!!options?.compat, {
+        blockHeight: String(value.blockHeight),
+        values: [/** @type string */ (value.value)],
+      });
+      yield value;
+    }
+  }
+
+  /**
+   * @param {Height} [maximum]
+   * @param {Height} [minimum]
+   * @param {boolean} [compatMode]
+   */
+  async function* reverseIterate(
+    maximum,
+    minimum,
+    compatMode = options?.compat,
+  ) {
```

Comment:
You might consider having separate implementations of `reverseIterate` that you call from here (`ReverseIterateCompatMode` vs `reverseIterateStreamMode`) depending on whether `compatMode` is in place, since the non-compat mode can be simpler and faster at runtime without having to lift value cells into stream cells. This might also lend clarity because “stream mode” never has to fall back to equality checks when walking backward into value cells. It can just stop with an error.